ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java

```diff
@@ -0,0 +1,215 @@
+    this.networkId = networkId;
+    this.networkInfo = networkInfo;
+    this.blockBroadcaster = blockBroadcaster;
+    this.protocolContext = protocolContext;
+    this.ethPeers = ethPeers;
+    this.syncState = syncState;
+    this.miningEnabled = miningEnabled;
+    this.transactionPool = transactionPool;
+  }
+
+  BlockStats convertToStats(final Block block, final Difficulty difficulty) {
+    return convertToStats(
+        block.getHeader(),
+        block.getBody().getTransactions(),
+        block.getBody().getOmmers(),
+        difficulty);
+  }
+
+  BlockStats convertToStats(
+      final BlockHeader header,
+      final List<Transaction> txs,
+      final List<BlockHeader> ommers,
+      final Difficulty difficulty) {
+    return new BlockStats(
+        UInt256.valueOf(header.getNumber()),
+        Hash.fromBytes(header.getHash()),
+        Hash.fromBytes(header.getParentHash()),
+        header.getTimestamp(),
+        Address.fromBytes(header.getCoinbase()),
+        header.getGasUsed(),
+        header.getGasLimit(),
+        header.getDifficulty().toUInt256(),
+        header.getDifficulty().add(difficulty).toUInt256(),
+        txs.stream().map(this::convertTxToStats).collect(Collectors.toList()),
+        Hash.fromBytes(header.getTransactionsRoot()),
+        Hash.fromBytes(header.getStateRoot()),
+        ommers.stream().map(this::convertToBlockHeader).collect(Collectors.toList()));
+  }
+
+  private org.apache.tuweni.eth.BlockHeader convertToBlockHeader(final BlockHeader header) {
+    return new org.apache.tuweni.eth.BlockHeader(
+        Hash.fromBytes(header.getParentHash()),
+        Hash.fromBytes(header.getOmmersHash()),
+        Address.fromBytes(header.getCoinbase()),
+        Hash.fromBytes(header.getStateRoot()),
+        Hash.fromBytes(header.getTransactionsRoot()),
+        Hash.fromBytes(header.getReceiptsRoot()),
+        header.getLogsBloom(),
+        header.getDifficulty().toUInt256(),
+        UInt256.valueOf(header.getNumber()),
+        Gas.valueOf(header.getGasLimit()),
+        Gas.valueOf(header.getGasUsed()),
+        Instant.ofEpochMilli(header.getTimestamp()),
+        header.getExtraData(),
+        Hash.fromBytes(header.getMixHash()),
+        Bytes.ofUnsignedLong(header.getNonce()));
+  }
+
+  private TxStats convertTxToStats(final Transaction transaction) {
+    return new TxStats(Hash.fromBytes(transaction.getHash()));
+  }
+
+  private void onNewBlock(final Block block, final Difficulty difficulty) {
+    reporter.sendNewHead(
+        convertToStats(
+            block.getHeader(),
+            block.getBody().getTransactions(),
+            block.getBody().getOmmers(),
+            difficulty));
+  }
+
+  public void start(final Vertx vertx) {
+    Log4j2LoggerProvider provider = new Log4j2LoggerProvider();
+    URI ethStatsURI =
+        URI.create(
+            "ws://" + ethStatsParameters.getHost() + ":" + ethStatsParameters.getPort() + "/api");
```

Comment:
i am not in the context of this code but using websocket without ssl seems something that might or not be important. just a  small warn.

Reply:
Good point. I don't remember how this is done on geth's side. I'll look.

Reply:
Guess what, they try wss first and default back to ws if wss fails.

Reply:
This is resolved now.